### ISSUE TITLE
VorbisComment - Adding new Tag (keyvalue pair)?

### ISSUE THREAD

**Issue by HomerJau** [AaronLenoir/flaclibsharp#43]
I'd like to add a new (non standard in flaclibsharp) tag to a FLAC file for storing replaygain data (requires multiple replaygain tags).

flaclibsharp can read those non-standard replaygain tags if they exist in a flac ok.

However there does not seem to be a way to create new tags or update non-standard tags.

Is it possible for you to add methods to:
Add a new key & value (example "replay_gain-track_gain=-2.50 db")
Update a non-standard tag  (example "replay_gain-track_gain=-2.50 db")

Question:
Will the existing VorbisComment .save method support saving new tags without trashing the file? (if I go ahead and modify flaclibsharp to allow new VorbisCommentvalues)

THX

**Comment by AaronLenoir** [comment]
Hello

Thanks for you interest! Hopefully we can find a solution for you.

I have some questions:

- Am I correct in assuming the replay_gain-track_gain are Vorbis Comment Tags? Or are we talking about something else, like custom FLAC metadata.
- Do you have an experience where the Save method for the VorbisComment trashes the file? Can you provide an example (perhaps in a seperate github issue)?

I believe adding (multiple) tags can be done as follows.

**However**: I just tried this and when reading I only get one of the values back for that tag! So I think there's a bug I will need to investigate that as soon as I have some time.

Have you tried the following?

```
            // Add multiple vorbis comment tags
            using (FlacFile file = new FlacFile(@"Data\testfile1.flac"))
            {
                if (file.VorbisComment == null)
                {
                    file.Metadata.Add(new VorbisComment());
                }

                file.VorbisComment["replay_gain-track_gain"] = new VorbisCommentValues(new string[] { "-2.50 db", "-3.50 db" });

                file.Save();
            }
```

**Comment by HomerJau** [comment]
Hi Aaron.

Thanks for getting back to me!!

Your Questions:
1. Yes – The ReplayGain tags (there will be 4 tags) are stored in the VorbisComment tags (I used MetaFlac to view these created with Foorbar2000’s ReplayGain add-in)
2. I did not try the save method. The ‘trash’ question  was just something I was wondering. I was thinking more data may write unwanted info into the file – maybe just a silly question)

I will try your suggested code below but the problem would be that there will almost always be existing VorbisComment tags so I will need a method to add the existing tags not just add new tags.

Here’s a dump of VorbisComent tags for a test file (there’s a bunch of MusicBrainz metadata too):

ALBUM=Explores Your Mind (Quad Q8)
ALBUMARTIST=Al Green
ARTIST=Al Green
DATE=1974
GENRE=Soul
TITLE=Take Me to the River
TRACKTOTAL=09
TRACKNUMBER=02
MUSICBRAINZ_ALBUMID=89c8aab3-f29b-4f96-be81-d333407118b3
MUSICBRAINZ_ALBUMSTATUS=Official
MUSICBRAINZ_ALBUMTYPE=Album
MUSICBRAINZ_ARTISTID=fb7272ba-f130-4f0a-934d-6eeea4c18c9a
MUSICBRAINZ_RELEASEGROUPID=ce3619a4-18a9-3dea-bc34-2e4cb9c8151d
MUSICBRAINZ_TRACKID=28f41205-4ed1-4ec6-bca2-8f77b36d726c
replaygain_album_gain=-7.64 dB
replaygain_album_peak=1.000000
replaygain_track_gain=-8.39 dB
replaygain_track_peak=1.000000


THX
Garry


Sent from Mail for Windows 10

**Comment by HomerJau** [comment]
Hi Aaron.

I have resolved my original question/issue.

After looking more closely at the FlacLibSharp VorbisComment class I could see the protected methods ‘AddComment’ add a new Tag Key/Value if it key doesn’t exist AND updates the key’s value if it does. No need to worry about existence or not.

I made changed one of these to a public method and now can add/edit new/existing tags with simple code from my app:

FlacFile file = new FlacFile(pFileName);
VorbisComment VorbisComment = file.VorbisComment;
if ((VorbisComment == null)) {
    file.Metadata.Add(new VorbisComment());
    VorbisComment = file.VorbisComment;
}
    
VorbisComment.AddComment("REPLAYGAIN_TRACK_GAIN", pTrackGain);
VorbisComment.AddComment("REPLAYGAIN_ALBUM_GAIN", pAlbumGain);
etc
 File.save()

(Maybe ‘AddComment’ should be named to ‘UpdateComment’to clarify its function?)


I do have one request: Please consider adding a ‘Remove’ to VorbisComments (removes a Tag Key/Value)
 

Thanks for creating FlacLibSharp. Works a treat!

**Comment by AaronLenoir** [comment]
Good to hear you got something to work. Sorry the API wasn't more clear. I made two issues to address the things we've found here: #44 and #45 
Feel free to add some comments there if I missed something. I may try to ask you some questions when I work on it.

I'll close this one since your original problem was solved.

I'm glad to hear you are finding this project useful. I'm curious to know what you're using it for, if you feel like sharing that.

Talk to you later!
